agent/src/com/thoughtworks/go/agent/service/AgentWebsocketService.java

```diff
@@ -0,0 +1,212 @@
+            service.send(new Message(Action.reportCurrentStatus, new Report(agentRuntimeInfo, jobIdentifier, jobState)));
+        }
+
+        @Override
+        public void reportCompleting(AgentRuntimeInfo agentRuntimeInfo, JobIdentifier jobIdentifier, JobResult result) {
+            service.send(new Message(Action.reportCompleting, new Report(agentRuntimeInfo, jobIdentifier, result)));
+        }
+
+        @Override
+        public void reportCompleted(AgentRuntimeInfo agentRuntimeInfo, JobIdentifier jobIdentifier, JobResult result) {
+            service.send(new Message(Action.reportCompleted, new Report(agentRuntimeInfo, jobIdentifier, result)));
+        }
+
+        @Override
+        public boolean isIgnored(JobIdentifier jobIdentifier) {
+            return runner.isJobCancelled();
+        }
+
+        @Override
+        public String getCookie(AgentIdentifier identifier, String location) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AgentWebsocketService.class);
+    private AgentController controller;
+    private Session session;
+    private URLService urlService;
+    private WebSocketClient client;
+    private Executor executor = Executors.newFixedThreadPool(5);
+
+    @Autowired
+    public AgentWebsocketService(URLService urlService) {
+        this.urlService = urlService;
+    }
+
+    public synchronized void start() throws Exception {
+        SslContextFactory sslContextFactory = new SslContextFactory();
+        sslContextFactory.setKeyStorePath(SslInfrastructureService.AGENT_CERTIFICATE_FILE.getAbsolutePath());
+        sslContextFactory.setKeyStorePassword(SslInfrastructureService.AGENT_STORE_PASSWORD);
+        sslContextFactory.setKeyManagerPassword(SslInfrastructureService.AGENT_STORE_PASSWORD);
+        sslContextFactory.setTrustStorePath(SslInfrastructureService.AGENT_TRUST_FILE.getAbsolutePath());
+        sslContextFactory.setTrustStorePassword(SslInfrastructureService.AGENT_STORE_PASSWORD);
+        sslContextFactory.setWantClientAuth(true);
+        if (client == null || client.isStopped()) {
+            SystemEnvironment environment = new SystemEnvironment();
+            client = new WebSocketClient(sslContextFactory);
+            client.setMaxIdleTimeout(environment.getWebsocketMaxIdleTime());
+            Message.setupPolicy(client.getPolicy(), environment);
+            client.start();
+        }
+        if (session != null) {
+            session.close();
+        }
+        LOGGER.info("Connecting to websocket endpoint: " + urlService.getAgentRemoteWebSocketUrl());
+        session = client.connect(this, new URI(urlService.getAgentRemoteWebSocketUrl()), new ClientUpgradeRequest()).get();
+    }
+
+    public synchronized void stop() {
+        if (isRunning()) {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("close " + sessionName());
```

Comment:
With SLF4j — this can be rewritten as `LOGGER.debug("close {}", sessionName())`. The actual string interpolation only happens if the message needs to be logged. You may drop the `isDebugEnabled` as well.